src/udev/udev-node.c

```diff
@@ -204,24 +206,111 @@ static int link_find_prioritized(sd_device *dev, bool add, const char *stackdir,
                         if (device_get_devlink_priority(tmp_dev, &tmp_prio) < 0)
                                 continue;
 
-                        if (target && tmp_prio <= priority)
+                        if (devnode && tmp_prio <= priority)
                                 continue;
 
-                        if (sd_device_get_devname(tmp_dev, &devnode) < 0)
+                        if (sd_device_get_devname(tmp_dev, &val) < 0)
                                 continue;
 
-                        r = free_and_strdup(&target, devnode);
+                        r = free_and_strdup(&devnode, val);
                         if (r < 0)
                                 return r;
-                }
+
+                } else
+                        continue;
 
                 priority = tmp_prio;
         }
 
-        *ret = TAKE_PTR(target);
+        *ret = TAKE_PTR(devnode);
         return !!*ret;
 }
 
+static int stack_directory_update(sd_device *dev, int fd, bool add) {
+        const char *id;
+        int r;
+
+        assert(dev);
+        assert(fd >= 0);
+
+        r = device_get_device_id(dev, &id);
+        if (r < 0)
+                return r;
+
+        if (add) {
+                _cleanup_free_ char *data = NULL, *buf = NULL;
+                const char *devname;
+                int priority;
+
+                r = sd_device_get_devname(dev, &devname);
+                if (r < 0)
+                        return r;
+
+                r = device_get_devlink_priority(dev, &priority);
+                if (r < 0)
+                        return r;
+
+                if (asprintf(&data, "%i:%s", priority, devname) < 0)
+                        return -ENOMEM;
+
+                if (readlinkat_malloc(fd, id, &buf) >= 0 && streq(buf, data))
+                        return 0; /* Unchanged. */
+
+                (void) unlinkat(fd, id, 0);
+
+                if (symlinkat(data, fd, id) < 0)
+                        return -errno;
+
+        } else {
+                if (unlinkat(fd, id, 0) < 0) {
+                        if (errno == ENOENT)
+                                return 0; /* Unchanged. */
+                        return -errno;
+                }
+        }
+
+        return 1; /* Updated. */
+}
+
+static int stack_directory_open(const char *dirname) {
+        _cleanup_close_ int fd = -1;
+        int r;
+
+        assert(dirname);
+
+        r = mkdir_parents(dirname, 0755);
+        if (r < 0)
+                return r;
+
+        fd = open_mkdir_at(AT_FDCWD, dirname, O_CLOEXEC | O_DIRECTORY | O_NOFOLLOW | O_RDONLY, 0700);
```

Comment:
this can be relaxed now that you use the lock file, no? i.e. 0755 to make things easier to debug